express-notes-api/server.js

```diff
@@ -0,0 +1,197 @@
+
+// Clients can GET a list of notes.
+async function readAllNotes() {
+  try {
+    const fileData = await readFile('data.json');
+    const data = JSON.parse(fileData);
+    const notes = Object.values(data.notes);
+    return notes;
+  } catch (error) {
+    console.error(`error: ${error}`);
+    throw error;
+  }
+}
+
+app.get('/api/notes', async (req, res) => {
+  try {
+    const notes = await readAllNotes();
+    res.status(200).json(notes);
+  } catch (error) {
+    res.sendStatus(500);
+  }
+});
+
+// Clients can GET a single note by id.
+app.get('/api/notes/:id', async (req, res) => {
+  try {
+    const id = parseInt(req.params.id);
+    if (isNaN(id) || id <= 0) {
+      res.status(400).json({ error: 'ID must be a positive integer.' });
+      return;
+    }
+
+    const fileData = await readFile('data.json');
+    const data = JSON.parse(fileData);
+    const note = data.notes[id];
+
+    if (note) {
+      res.status(200).json(note);
+    } else {
+      res.status(404).json({ error: `Cannot find note with id ${id}` });
+    }
+  } catch (error) {
+    console.log(`error: ${error}`);
+    res.sendStatus(500);
+  }
+});
+
+// Clients can POST a new note.
+let nextId = 1;
+
+async function createNote(notes) {
+  try {
+    const data = {
+      notes: {},
+    };
+
+    notes.forEach((note) => {
+      data.notes[note.id] = note;
+    });
+
+    data.nextId++;
+
+    await writeFile('data.json', JSON.stringify(data, undefined, 2));
+  } catch (error) {
+    console.error(`error: ${error}`);
+    throw error;
+  }
+}
+
+app.use(express.json());
+
+app.post('/api/notes', async (req, res) => {
+  try {
+    const content = req.body.content;
+    if (!content) {
+      res.status(400).json({ error: 'Content is required.' });
+      return;
+    }
+
+    const notes = await readAllNotes();
+    const newNote = {
+      id: nextId++,
+      content,
+    };
+
+    notes.push(newNote);
+    await createNote(notes);
+
+    res.status(201).json(newNote);
+  } catch (error) {
+    console.error(`Error: ${error}`);
+    res.status(500).json({ error: 'An unexpected error occurred.' });
+  }
+});
+
+// Clients can DELETE a note by id.
+async function deleteNote(noteId) {
+  try {
+    const fileData = await readFile('data.json');
+    const data = JSON.parse(fileData);
+
+    if (!(noteId in data.notes)) {
+      const error = new Error(`Cannot find note with id ${noteId}`);
+      error.notFound = true;
+      throw error;
+    }
+
+    delete data.notes[noteId];
+
+    const updatedData = JSON.stringify(data, undefined, 2);
+
+    await writeFile('data.json', updatedData);
+  } catch (error) {
+    console.log(`error: ${error}`);
+    throw error;
+  }
+}
+
+app.delete('/api/notes/:id', async (req, res) => {
+  try {
+    const id = parseInt(req.params.id);
+    if (isNaN(id) || id <= 0) {
+      res.status(400).json({ error: 'ID must be a positive integer.' });
+      return;
+    }
+
+    try {
+      await deleteNote(id);
```

Comment:
I like how you have extracted all the primary functionality for your handlers into separate functions. It would make it even better if you put all those into a separate file. As is, it's a bit challenging to read through your code since it jumps from handler to implementation and back again.